### PR TITLE
Improve arc doc, fixing #32905

### PR DESCRIPTION
**Pull request by izgzhen**
As issue #32905 detailed, I moved part of the module doc to the struct doc, and fixed some small places in the `alloc::arc`.

### CODE REVIEW

File: src/liballoc/arc.rs [hunk 1]
```diff
@@ -92,15 +59,23 @@ use heap::deallocate;
...
+/// the [`Rc<T>` type](../rc/struct.Rc.html). It is the same as `Arc<T>`, but
+/// does not use atomics, making it both thread-unsafe as well as significantly
+/// faster when updating the reference count.
 ///
 /// # Examples
 ///
-/// In this example, a large vector is shared between several threads.
-/// With simple pipes, without `Arc`, a copy would have to be made for each
+/// In this example, a large vector of data will be used by several threads.
+/// With simple pipes (XXX: wat?) instead of `Arc`, a copy would have to be made for each
```
Comment (izgzhen) [inline src/liballoc/arc.rs]: Not clear what does `pipes` really mean here, any comments?

**Comment by izgzhen** [comment]
Note that I marked the `pipes` in the diff and had some problem in how to handling this.

File: src/liballoc/arc.rs (hunk 1 above)
Comment (steveklabnik) [inline src/liballoc/arc.rs]: I think it's referring to channels, but we can probably just remove the whole sentence.

Comment (izgzhen) [inline src/liballoc/arc.rs]: Removing looks good to me. It might be a distraction I suppose.

**Comment by brson** [comment]
lgtm when @steveklabnik's comments addressed

File: src/liballoc/arc.rs [hunk 2]
```diff
@@ -92,15 +59,19 @@ use heap::deallocate;
...
-/// In this example, a large vector is shared between several threads.
-/// With simple pipes, without `Arc`, a copy would have to be made for each
-/// thread.
+/// # Examples
 ///
-/// When you clone an `Arc<T>`, it will create another pointer to the data and
-/// increase the reference counter.
+/// In this example, a large vector of data will be shared by several threads. First we
+/// wrap it with a `Arc::new` and then clone the `Arc<T>` reference for every thread (which will
+/// increase the reference count atomically).
```
Comment (izgzhen) [inline src/liballoc/arc.rs]: @steveklabnik How about this?

**Comment by steveklabnik** [comment]
Looks good, thanks a ton!

@bors: r+ rollup

**Pull request merged**